src/local/cost_matrix.rs

```diff
@@ -21,21 +25,39 @@ impl LocalCostMatrix {
     #[inline]
     pub fn new() -> Self {
         LocalCostMatrix {
-            bits: vec![0; 2500],
+            bits: [0; 2500],
         }
     }
 
     #[inline]
     pub fn set(&mut self, x: u8, y: u8, val: u8) {
-        self.bits[pos_as_idx(x, y)] = val;
+        self[(x, y)] = val;
     }
 
     #[inline]
     pub fn get(&self, x: u8, y: u8) -> u8 {
-        self.bits[pos_as_idx(x, y)]
+        self[(x, y)]
+    }
+
+    // # Safety
+    // Calling this method with x >= 50 or y >= 50 is undefined behaviour.
+    #[inline]
+    pub unsafe fn get_unchecked(&self, x: u8, y: u8) -> u8 {
+        debug_assert!(x < 50, "out of bounds x: {}", x);
```

Comment:
Might be simpler to add a function to get the reference to a entry in the cost matrix. That way you can move the checks and the unchecked get in to a single location versus duplicating.

pub unsafe fn get_entry(&mut self, x: u8, y: u8) -> &mut u8 {
    debug_assert!(x < 50, "out of bounds x: {}", x);
    debug_assert!(y < 50, "out of bounds y: {}", y);
    self.bits.get_unchecked_mut(pos_as_idx(x, y));
}

Reply:
I was thinking about having a `get_unchecked_mut`, but even then you'd need both due to `&self` vs `&mut self`.

Reply:
You're right - you'd still need both. It would still reduce the number of places the asserts are needed from 4 down to too so still worth it.

Reply:
I'm still not sure how we'd get down to 2; there's the non-debug asserts in the indexing impls, which are there for correctness, and there's the debug asserts in the unchecked methods, which are there as testing aids. Can't replace one sort with the other, since they're for different purposes.

Reply:
I had missed the nuance on the asset vs debug_assert - you're using the debug_asset to reduce the cost in places where it shouldn't panic. The raw u8 indexing could come from user values and as such needs to be untrusted and panic.

The only way to factor it all out so there's 2 places using get_unchecked(_mut) would end up with the assert functions calling to ones that have debug_asset which is a bit gross. You could still combine the position indexing with the get/set functions.

Reply:
You mean have the indexing functions call in to the self.get_unchecked? That would work if it returned a reference, which I suppose it could do. Would be weird having get returning u8 and get_unchecked returning &u8 (similar with set vs get_unchecked_mut).